contributors/design-proposals/secure_watch.md

```diff
@@ -0,0 +1,208 @@
+```
+
+ The NodeSelector field will be added to ```ListOptions ``` (next to label & field
+selectors) and will be supported only by LIST and WATCH requests.
+
+ With that mechanism in place, all List/Watch requests coming from kubelets will have
+to have this field correctly set. We will create a dedicated admission plugin that will
+be responsible for checking if a given ```node selector ``` is allowed from a given
+client (the exact mechanism for doing this is out of scope for this doc) and either
+rejecting this request or letting it go. Note that doing this may require modifying
+admission attributes from implementation point of view.
+
+ TODO: Consider adding NodeSelector to ```GetOptions ``` - if we would do that,
+we could have unified patern from authorizing all requests from nodes and have the
+admission plugin to be relatively simple.
+
+
+### Semantic of watch
+
+ To make all our existing "list-watch-related" frameworks work correctly for this
+kind of watch, we would like to preserve all the crucial invariants of watch. This
+in particular means:
+
+1. There is at most one watch event for a given resource version.
+2. Watch events are delivered in the increasing order of resource versions.
+
+ In the ideal situation, if a new pod that is referencing object X appears (and
+the we are watching objects of that type), we send add event for object X.
+However, that would break the above assumptions, because:
+
+1. There can be more objects referenced by a given pod (so we can't send all of
+them with a rv corresponding to that pod add/update/delete)
+2. If we decide for sending those with their original rv`s, then we could
+potentially go back in time.
+
+ As a result, we propose the following semantic:
+
+1. If a pod is being created/updated/deleted, no watch events are delivered.
+It's responsibility of a user to grab current version of all objects that are
+being referenced by this pods.
+2. From that point, events for all objects being referenced by this pod (e.g.
+add/update/delete of secret being referenced by this pod) are being delivered
+to the watcher as long as the pod exists on the node.
+
+ With this semantic we will be able to reuse e.g. reflector framework with just
+minor modifications.
+
+TODO: Describe how this should be used.
+
+
+### Determining whether object is referenced by pods from a given node.
+
+ The tricky part in determining whether an object X is referenced by any pods
+bound to a given node is to avoid different kind of race conditions and do it
+in deterministic way.
+
+ The crucial requirements here are:
+
+1. Whenever "list" request returns a list of objects and a resource version "rv",
+starting a watch from the returned "rv" will never drop any events.
+2. For a given watch request (with resource version "rv"), the returned stream
+of events is always the same (e.g. very slow laggin watch may not cause dropped
+events). 
+
+ To satisfy the above requirements, we can't really rely only on the existing
```

Comment:
This is going to be really hard to get right, I think. An alternative that comes to mind:

1. When the controller adds or removes a permission to view an object, it "touches" the object, e.g. increments a counter in a new field. 
2. Then the normal filtering mechanism will work with no changes whatsoever.